File: CustomViewObjc/CustomViewTemplate.m
//
//  CustomViewTemplate.m

//
//  Created by APP559 on 2014/7/28.
//  Copyright (c) 2014年 yume. All rights reserved.
//

#import "CustomViewTemplate.h"
#import "ViewSource.h"

@implementation CustomViewTemplate

- (instancetype)initWithCoder:(NSCoder *)aDecoder{
    self = [super initWithCoder:aDecoder];
    
    if (self) {
        [self setup];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    
    if (self) {
        [self setup];
    }
    return self;
}

- (void) setup{
    NSString *nibName = NSStringFromClass([self class]);
    
    UINib *nib = [UINib nibWithNibName:nibName bundle:[self frameworkBundle]];
    
    [nib instantiateWithOwner:self options:nil];
}

-(void)prepareForInterfaceBuilder{
    [self viewLiveRendering];
}


- (void)drawRect:(CGRect)rect{
#ifndef TARGET_INTERFACE_BUILDER
    [self viewLiveRendering];
#endif
    
    self.layer.borderColor = _borderColor.CGColor;
    self.layer.borderWidth = _borderLineWidth;
    
    if( [self.layer respondsToSelector:@selector(setCornerRadius:)] )
        [self.layer setCornerRadius:_borderRadius];
}

-(void)viewLiveRendering{
    
    if ([self respondsToSelector:@selector(processViewSource)]) {
        if (_viewSourceKeyPath) {
            _viewSourceDictionary =  [ViewSourceInstance valueForKey:_viewSourceKeyPath];
            
            NSString *type = _viewSourceDictionary[@"type"];
            NSString *className = NSStringFromClass([self class]);
            
            if ([className isEqualToString:type]) {
                [self processViewSource];
            }
        }
    }
    
    if ([self respondsToSelector:@selector(processStoryBoardSource)]) {
        [self processStoryBoardSource];
    }
    
    if ([self respondsToSelector:@selector(processFuture)]) {
        [self processFuture];
    }
    
}

- (NSBundle *)frameworkBundle{
    return [NSBundle bundleWithIdentifier:[self bundleIdentifier]];
}

- (NSString *)bundleIdentifier{
    //Bundle Identifier can be find at Target -> Your Framework -> Bundle Identifier
    return @"com.yume190.CustomViewObjc";
}

-(id)debugQuickLookObject{
    return self;
}

@end

File: CustomViewSwift/CustomViewTemplate.swift
//
//  CustomViewTemplate.swift
//  APSMSwift
//
//  Created by Yume on 2014/8/3.
//  Copyright (c) 2014年 yume. All rights reserved.
//

import UIKit

protocol CustomViewTemplateProtocol{
    func setup()
    func viewLiveRendering()
    
    func processViewSource()
    func processFuture()
    
    func className() -> String
}

protocol CustomViewProtocol:CustomViewTemplateProtocol{
    func instantiateWithXib()
    func bundleIdentifier() -> String
    func frameworkBundle() -> NSBundle
}

@IBDesignable
class CustomViewTemplate: UIView ,CustomViewTemplateProtocol{
    
    @IBInspectable var viewSourceKeyPath:String?
    @IBInspectable var viewSourceDictionary:NSDictionary?
    
    @IBInspectable var borderLineWidth: CGFloat = 0 {
        willSet{
            layer.borderWidth = newValue
        }
    }
    
    @IBInspectable var borderRadius: CGFloat = 0 {
        willSet{
            layer.cornerRadius = newValue
        }
    }
    
    @IBInspectable var borderColor: UIColor = UIColor.clearColor() {
        willSet{
            layer.borderColor = newValue.CGColor
        }
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init(coder : NSCoder) {
        super.init(coder:coder)
        setup()
    }
    
    func setup() {}
    
    override func prepareForInterfaceBuilder() {
        self.viewLiveRendering()
    }
    
    override func drawRect(rect: CGRect) {
        self.viewLiveRendering()
    }
    
    func viewLiveRendering() {
        if respondsToSelector(Selector("processViewSource")){
            if viewSourceKeyPath != nil{
                viewSourceDictionary = ViewSourceInstance.valueForKeyPath(viewSourceKeyPath!) as? NSDictionary
                
                let type:NSString? = viewSourceDictionary!["type"] as? NSString
                
                if type!.isEqualToString(className()) {
                    processViewSource()
                }
            }
        }
        
        if respondsToSelector(Selector("processFuture")){
            processFuture()
        }
    }
    
    func processViewSource() {}
    func processFuture() {}
    
    func instantiateWithXib(){
        
        var bundle:NSBundle = frameworkBundle()
        
        var nib:UINib = UINib(nibName: className(), bundle: bundle)
        
        nib.instantiateWithOwner(self, options: nil)
    }
    
    func frameworkBundle() -> NSBundle{
        return NSBundle(identifier: bundleIdentifier())!
    }
    
    func bundleIdentifier() -> String{
        //Bundle Identifier can be find at Target -> Your Framework -> Bundle Identifier
        return "com.yume190.CustomViewSwift"
    }
    
    func className() -> String{
        return ""
    }
}
